src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs

```diff
@@ -0,0 +1,174 @@
+using Microsoft.Build.Framework;
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Net;
+using System.Collections.Generic;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public sealed class DownloadFilesFromUrl : BuildTask
+    {
+        private static readonly HttpClient s_client = new HttpClient(GetHttpHandler());
+
+        /// <summary>
+        /// The items to download.
+        /// Url and DestinationFile are required in the item's metadata, DestinationDir is optional.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        /// <summary>
+        /// The destination directory for all the items to be downloaded.
+        /// This parameter can be overriden per item, if an item has DestinationDir in it's metadata, we will take that directory.
+        /// If both, the item's and this property are empty, the items will be downloaded to the working directory.
+        /// </summary>
+        public string DestinationDir { get; set; }
+
+        /// <summary>
+        /// The default is to fail on error, set this to true if you want to just warn.
+        /// </summary>
+        public bool TreatErrorsAsWarnings { get; set; }
+
+        /// <summary>
+        /// The list of files created. It is not guaranted that all the input items will be successfully downloaded
+        /// when TreatErrorsAsWarings is set to true.
+        /// </summary>
+        [Output]
+        public ITaskItem[] FilesCreated { get; set; }
+
+        public override bool Execute()
+        {
+            var filesCreated = new List<ITaskItem>();
+            foreach (var item in Items)
+            {
+                string downloadSource = item.GetMetadata("Url");
+                if (!Uri.IsWellFormedUriString(downloadSource, UriKind.Absolute))
+                {
+                    if (TreatErrorsAsWarnings)
+                    {
+                        Log.LogWarning($"Item {item.ItemSpec} Url is not a valid url.");
+                        continue;
+                    }
+                    else
+                    {
+                        return ExitWithError($"Item {item.ItemSpec} Url is not a valid url.");
+                    }
+                }
+
+                Uri downloadUri = new Uri(downloadSource);
+                string fileName = item.GetMetadata("DestinationFile");
+                if (string.IsNullOrWhiteSpace(fileName))
+                {
+                    // starts with file://
+                    if (downloadUri.IsFile)
+                    {
+                        fileName = Path.GetFileName(downloadUri.LocalPath);
```

Comment:
I think you can actually use Path.GetFileName on both cases instead of trying to parse the Uri yourself below.

Reply:
Ok, will try that. In that case we wouldn't need the downloadUri.IsFile check. Will try it.

Reply:
Fixed. Thanks for the feedback.